### ISSUE TITLE
Exchanging Google ID Token for Keycloak Access Token results in Error with reason "subject_token_type invalid"

### ISSUE THREAD

**Issue by kumaraish** [keycloak/keycloak#20042]
### Before reporting an issue

- [X] I have searched existing issues
- [X] I have reproduced the issue with the latest release

### Area

token-exchange

### Describe the bug

Following error is seen in keycloak logs when an external to internal token-exchange request is made to exchange Google ID Token with Keycloak Access Token

`2023-04-30 12:53:04,984 WARN  [org.keycloak.events] (executor-thread-22) type=TOKEN_EXCHANGE_ERROR, realmId=9edbb8fd-a6cc-4cf3-9632-4293bdc76f3f, clientId=myclient, userId=null, ipAddress=172.17.0.1, error=invalid_token_type, reason='subject_token_type invalid', auth_method=token_exchange, grant_type=urn:ietf:params:oauth:grant-type:token-exchange, subject_issuer=google, client_auth_method=client-secret`

### Version

21.1.0

### Expected behavior

As per the documentation here https://www.keycloak.org/docs/latest/securing_apps/#external-token-to-internal-token-exchange - 
- a new user is imported into your realm if it doesn’t exist
- When the exchange is complete, a user session will be created within the realm, and you will receive an access and or refresh token depending on the requested_token_type parameter value

Based on the above, expectation is that Keycloak should create an user based on the user information in Google ID token and should return an access token

### Actual behavior

Keycloak returning "HTTP 400 Bad Request"  with this error {"error":"invalid_token","error_description":"invalid token type"} 

Error in KeyCloak logs

2023-04-30 12:53:04,984 WARN  [org.keycloak.events] (executor-thread-22) type=TOKEN_EXCHANGE_ERROR, realmId=9edbb8fd-a6cc-4cf3-9632-4293bdc76f3f, clientId=myclient, userId=null, ipAddress=172.17.0.1, error=invalid_token_type, reason='subject_token_type invalid', auth_method=token_exchange, grant_type=urn:ietf:params:oauth:grant-type:token-exchange, subject_issuer=google, client_auth_method=client-secret

### How to Reproduce?

1. Start Keycloak with `--features=token-exchange,admin-fine-grained-authz` enabled
2. Configure Google as Identity Provider in Keycloak
3. Get a Google ID Token (outside keycloak)
4. Post a request to keycloak to exchange this ID token with a Keycloak internal access token

```
export id_token=eyJhbGciOiJSUzI1N...

curl -v -X POST http://localhost:8180/realms/habits/protocol/openid-connect/token \
--data-urlencode "client_id=myclient" \
--data-urlencode "grant_type=urn:ietf:params:oauth:grant-type:token-exchange" \
--data-urlencode "subject_token=$id_token" \
--data-urlencode "subject_issuer=google" \
--data-urlencode "requested_token_type=urn:ietf:params:oauth:token-type:access_token" \
--data-urlencode "subject_token_type=urn:ietf:params:oauth:token-type:id_token"
```

### Anything else?

Could this be the probable cause?

https://github.com/keycloak/keycloak/blob/3f6925143aa8a3e61c30782029ca5ea781d60ce0/services/src/main/java/org/keycloak/social/google/GoogleIdentityProvider.java#LL92C16-L92C54

```java
@Override
protected BrokeredIdentityContext exchangeExternalImpl(EventBuilder event, MultivaluedMap<String, String> params) {
    return exchangeExternalUserInfoValidationOnly(event, params);
}
```

The linked function returns `exchangeExternalUserInfoValidationOnly` which is defined in `AbstractOAuth2IdentityProvider.java` and expects only ACCESS_TOKEN

```java
@Override
protected BrokeredIdentityContext exchangeExternalImpl(EventBuilder event, MultivaluedMap<String, String> params) {
    return exchangeExternalImpl(event, params);
}
```

While `exchangeExternalImpl` defined in `OIDCIdentityProvider` (which `GoogleIdentityProvider` extends) validates id_token and JWT as well.

**Comment by sschu** [comment]
The documentation says: ```
The subject_token_type must either be urn:ietf:params:oauth:token-type:access_token or urn:ietf:params:oauth:token-type:jwt. If the type is urn:ietf:params:oauth:token-type:access_token you specify the subject_issuer parameter and it must be the alias of the configured identity provider. If the type is urn:ietf:params:oauth:token-type:jwt, the provider will be matched via the issuer claim within the JWT which must be the alias of the provider, or a registered issuer within the providers configuration.
```
So `subject_token_type=urn:ietf:params:oauth:token-type:id_token` is not valid.

**Comment by kumaraish** [comment]
And `urn:ietf:params:oauth:token-type:jwt` doesn't works either.

**Comment by sschu** [comment]
I agree with your analysis of the cause. However, I cannot tell you whether this was intended or not.

**Comment by kumaraish** [comment]
What's the way forward then? Clearly, my use case is not working out because of this issue.

**Comment by sschu** [comment]
I'll try to find out what is the intended behaviour here...

**Comment by mposolda** [comment]
Thanks for the report, but unfortunately due the amount of other reported issues and other priorities, Keycloak team does not have time to properly triage this bug. So preliminary added to Backlog for now.
It will be helpful if:

- You can verify if still applicable in latest Keycloak released version. If not, then it is welcome to close this issue.
- If you figure that this may not be a valid bug (for example just a mistake in configuration etc), it is also welcome to close this issue
- If still applicable in latest version, it will be welcome to add the comment as well, that this was still reproduced with latest Keycloak version as it is very valuable info. Anyone is welcome to comment with this or add other relevant comments to this issue.
- If you figure this is valid bug and you know how to fix it, you're welcome to contribute the fix for this. See the "Contributors" section here https://www.keycloak.org/community for the details on how to contribute to Keycloak and eventually send the PR.

**Comment by albilaga** [comment]
I also have this issue. When using `token_exchange` with `subject_token_type` = `urn:ietf:params:oauth:token-type:access_token` with google, keycloak request `access_token` instead of `id_token`(jwt)
The problem is the new [CredentialManager from Google](https://developer.android.com/training/sign-in/credential-manager) is only giving `id_token` (not really sure if we can get access token.

When I try to use `subject_token_type` = `urn:ietf:params:oauth:token-type:jwt` then it will also invalid because issuer from jwt from google is `https://accounts.google.com` instead my google alias even though I set `subject_issuer` = `google`

**Comment by mposolda** [comment]
Adding to backlog for now. We may revisit once we work on supported token exchange feature. In the meantime, the PR (including automated test) is welcome as long as this issue is real Keycloak issue. See the details about how to contribute PR https://github.com/keycloak/keycloak/blob/main/CONTRIBUTING.md

**Comment by mposolda** [comment]
@albilaga @kumaraish I've recently tried Google and it returns both ID-token and access-token from the normal social login. Considering this, I am able to use Google access token.

Could someone of you (or someone else) please confirm if it is still needed to exchange Google ID-tokens to Keycloak access-tokens? Or is it sufficient to rely on exchanging Google access-tokens (which means it may not be needed anymore to support exchanging Google ID-tokens) ?

**Comment by albilaga** [comment]
@mposolda I think `Google ID Token` still needed. If the client want to build native app (that don't throw the user to login to google into the web) then for android we have to use [Sign In With Google](https://developer.android.com/identity/sign-in/credential-manager-siwg) which only return Google Id Token

**Comment by AlexSilva-dev** [comment]
@mposolda @albilaga 
it is my case, i need for my native app.
I'm looking for the native experience but Keycloak doesn't support it :(

**Comment by Emil-Ey** [comment]
@mposolda 
We are using Keycloak Token-Exchange for authenticating through Google using native login. This will be broken once we [Migrate from legacy Google Sign-In to Credential Manager and AuthorizationClient](https://developer.android.com/identity/sign-in/legacy-gsi-migration) in our Android client. 

Google has deprecated Google Sign-In for Android: [Google blog post](https://android-developers.googleblog.com/2024/09/streamlining-android-authentication-credential-manager-replaces-legacy-apis.html) and will remove it from the SDK in H2 2025 (i.e. any moment). 

The new way to [Authenticate users with Sign in with Google](https://developer.android.com/identity/sign-in/credential-manager-siwg) will only provide the `Google ID Token`, and not the old `access_token`. Thus we will not be able to use Keycloak token exchange to authenticate Android users using Google as IDP if this bug/feature is not fixed/implemented.

**Comment by ragnorak-dev** [comment]
I'm in the same situation, I'm integrating google signin on my app has new feature, and the best way is integrate with new Google Signin system (Credentials), but keycloak doesn't support it...

**Comment by chornthorn** [comment]
Do you guys have any update on this issue? I face the same issue now.

**Comment by dominikweifieg** [comment]
@chornthorn @ragnorak-dev @Emil-Ey @AlexSilva-dev   A solution for native apps can be achieved by fetching the id_token from the Google Credential Manager and then exchanging the id token into an access token using AuthorizationClient. This access token can then be exchanged by Keycloak. So, just one step more, but also one more dependency and one more network request.

```kotlin
// Step 2: Get Google access token via AuthorizationClient
val activity = context as Activity
val authRequest = AuthorizationRequest.builder()
    .setRequestedScopes(listOf(
        Scope("openid"),
        Scope("email"),
        Scope("profile"),
    ))
    .build()

val authResult = Identity.getAuthorizationClient(activity)
    .authorize(authRequest)
    .await()

val googleAccessToken = authResult.accessToken
    ?: throw Exception("No access token received from Google")
```
